src/main/host/syscall/socket.rs

```diff
@@ -0,0 +1,578 @@
+use crate::cshadow as c;
+use crate::host::context::{ThreadContext, ThreadContextObjs};
+use crate::host::descriptor::socket::unix::{UnixSocketFile, UnixSocketType};
+use crate::host::descriptor::socket::SocketFile;
+use crate::host::descriptor::{
+    CompatDescriptor, Descriptor, DescriptorFlags, FileMode, FileState, FileStatus, PosixFile,
+};
+use crate::host::process::Process;
+use crate::host::syscall::{self, Trigger};
+use crate::host::syscall_condition::SysCallCondition;
+use crate::host::syscall_types::{PluginPtr, SysCallArgs, TypedPluginPtr};
+use crate::host::syscall_types::{SyscallError, SyscallResult};
+use crate::utility::event_queue::EventQueue;
+
+use std::convert::{TryFrom, TryInto};
+use std::sync::Arc;
+
+use log::*;
+use nix::errno::Errno;
+use nix::sys::socket::MsgFlags;
+
+pub fn sendto(ctx: &mut ThreadContext, args: &SysCallArgs) -> SyscallResult {
+    let fd: libc::c_int = args.get(0).into();
+    let buf_ptr: PluginPtr = args.get(1).into();
+    let buf_len: libc::size_t = args.get(2).into();
+    let flags: libc::c_int = args.get(3).into();
+    let addr_ptr: PluginPtr = args.get(4).into();
+    let addr_len: libc::socklen_t = args.get(5).into();
+
+    // get the descriptor, or return early if it doesn't exist
+    let desc = match syscall::get_descriptor(ctx.process, fd)? {
+        CompatDescriptor::New(desc) => desc,
+        // if it's a legacy descriptor, use the C syscall handler instead
+        CompatDescriptor::Legacy(_) => {
+            return unsafe {
+                c::syscallhandler_sendto(
+                    ctx.thread.csyscallhandler(),
+                    args as *const c::SysCallArgs,
+                )
+                .into()
+            }
+        }
+    };
+
+    let posix_file = desc.get_file().clone();
+
+    // get the socket for the descriptor
+    let socket = match posix_file {
+        PosixFile::Socket(ref x) => x,
+        _ => return Err(Errno::ENOTSOCK.into()),
+    };
+
+    sendto_helper(ctx, socket, buf_ptr, buf_len, flags, addr_ptr, addr_len)
+}
+
+pub fn sendto_helper(
+    ctx: &mut ThreadContext,
+    socket: &SocketFile,
+    buf_ptr: PluginPtr,
+    buf_len: libc::size_t,
+    flags: libc::c_int,
+    addr_ptr: PluginPtr,
+    addr_len: libc::socklen_t,
+) -> SyscallResult {
+    // get the send flags
+    let flags = match MsgFlags::from_bits(flags) {
+        Some(x) => x,
+        None => {
+            warn!("Invalid sendto flags: {}", flags);
+            MsgFlags::from_bits_truncate(flags)
```

Comment:
Is there a reason not to return EINVAL for this case?

Reply:
Could you add a comment explaining that?

Reply:
Added.